### ISSUE TITLE
Exception when concurrent threads are trying to build the same Mapper

### ISSUE THREAD

**Issue by blacksmith77** [jmapper-framework/jmapper-core#29]
Ciao Alessandro!

I noticed this exception when 2 threads are trying to build the same Mapper and no mapper are available on cache (eg: app just started)

> com.googlecode.jmapper.exceptions.JMapperException: javassist.CannotCompileException: by java.lang.LinkageError: loader (instance of  org/apache/catalina/loader/WebappClassLoader): attempted  duplicate class definition for name: "comerawebgroupshakeThePAWcorelogicboBuddycomerawebgroupshakeThePAWcorelogicdaojpapoBuddyPO" 
>     at com.googlecode.jmapper.config.JmapperLog.ERROR(JmapperLog.java:46)
>     at com.googlecode.jmapper.JMapper.<init>(JMapper.java:418)
>     at com.googlecode.jmapper.JMapper.<init>(JMapper.java:358)
>     at com.erawebgroup.shakeThePAW.core.logic.dao.converters.BuddyPOConverter.convertPOtoBO(BuddyPOConverter.java:95)
>     at com.erawebgroup.shakeThePAW.core.logic.dao.jpa.STPJPABuddyDAO.findByEmail(STPJPABuddyDAO.java:330)
>     at com.developerscrappad.rest.business.RESTResourceNotification.getMyPendingNotificationsCount(RESTResourceNotification.java:141)

... when the mapper is available in cache, no exceptions are rised. (eg: a second call of the same function, even from concurrent threads)

Maybe this behavior is simply "working as intended".
I solved this "non problem" by triggering Mappers caching at app  start.

BTW, the chance that concurrent threads trigger the build of same mapper, in general, is very low ;-)

**Personal note:**
Grazie per questo fantastico contributo alla community!
A presto!

**Comment by avurro** [comment]
Hi friend!

this issue is solved with version 1.5, updates without worrying, the versions are backward compatible.

**Personal note:**
Ti ringrazio per il supporto! aggiungi una stella al progetto se ti va :)

**Comment by blacksmith77** [comment]
ouch, sorry for the "noob" report :-/
I forgot to mention that I upgraded to jmapper-core 1.6.0.
I update pom.xml and triggered Maven update of dependencies from Eclipse.

>	<dependency>
>     	<groupId>com.googlecode.jmapper-framework</groupId>
>     	<artifactId>jmapper-core</artifactId>
>     	<version>1.6.0</version>
> 	</dependency>

Am I missing something?

**Comment by avurro** [comment]
The principal method is synchronized:
```java
 private synchronized IMapper<D,S> createMapper(MapperBuilder mapper) ...
```
It is really strange that you will have this problem.
Can you supply the code to be tested?

**Comment by blacksmith77** [comment]
Hi!

I prepared an extraction of involved entites.
It's attached to this comment.

By executing JMapperTester I replicate the concurrent threads situation. 

> SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
> SLF4J: Defaulting to no-operation (NOP) logger implementation
> SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
> Exception in thread "Thread-1" com.googlecode.jmapper.exceptions.JMapperException: javassist.CannotCompileException: by java.lang.LinkageError: loader (instance of  sun/misc/Launcher$AppClassLoader): attempted  duplicate class definition for name: "comblacksmithcrewjmappertesterBuddycomblacksmithcrewjmappertesterBuddyPO"
> 	at com.googlecode.jmapper.config.JmapperLog.ERROR(JmapperLog.java:46)
> 	at com.googlecode.jmapper.JMapper.<init>(JMapper.java:445)
> 	at com.googlecode.jmapper.JMapper.<init>(JMapper.java:372)
> 	at com.googlecode.jmapper.JMapper.<init>(JMapper.java:359)
> 	at com.blacksmithcrew.jmappertester.BuddyPOConverter.convertPOtoBO(BuddyPOConverter.java:91)
> 	at com.blacksmithcrew.jmappertester.JMapperTester.run(JMapperTester.java:23)
> 	at java.lang.Thread.run(Thread.java:745)

[JMapperTester.jar.zip](https://github.com/jmapper-framework/jmapper-core/files/204555/JMapperTester.jar.zip)

**Comment by avurro** [comment]
Fixed! Thank you for reporting this problem and the support!

Just two considerations on your code:
```java
// Map ID (JMapper doesn't work for this)
bo.setId( po.getId() );
```
It's false, JMapper manages this case, i have tried and it works.

You can use `@JGlobalMap` instead of a long series of `@JMap` annotations.
```java
@JGlobalMap(excluded={"locale","gender"})
public class BuddyPO {
  ...
}
```

This fix will be released with 1.6.1 version! thank you again!
